shared/subspace-data-retrieval/src/piece_fetcher.rs

```diff
@@ -14,5 +14,92 @@
 // limitations under the License.
 
 //! Fetching pieces of the archived history of Subspace Network.
-//!
-//! TODO: move piece fetching here
+
+use crate::piece_getter::{ObjectPieceGetter, PieceGetterError};
+use futures::stream::FuturesUnordered;
+use futures::StreamExt;
+use subspace_core_primitives::{Piece, PieceIndex};
+use tracing::{debug, trace};
+
+/// Concurrently downloads the exact pieces in `piece_indexes`, returning them in that order.
+/// Each piece index must be unique.
+///
+/// If any piece can't be downloaded, returns an error.
+// This code was copied and modified from subspace_service::sync_from_dsn::download_and_reconstruct_blocks():
+// <https://github.com/autonomys/subspace/blob/d71ca47e45e1b53cd2e472413caa23472a91cd74/crates/subspace-service/src/sync_from_dsn/import_blocks.rs#L236-L322>
+pub async fn download_pieces<PG>(
+    piece_indexes: impl IntoIterator<Item = PieceIndex>,
+    piece_getter: &PG,
+) -> Result<Vec<Piece>, PieceGetterError>
+where
+    PG: ObjectPieceGetter,
+{
+    let piece_indexes = piece_indexes.into_iter().collect::<Vec<_>>();
```

Comment:
What is the point of having `impl IntoIterator` just to call `.into_iter().collect::<Vec<_>>()` on it and then `piece_indexes.iter()` below yet again?

Just make the argument either a slice of indices or `ExactSizeIterator` so you can avoid this mess.

I see that the goal is to collect it just so you can log them, but maybe you should log them at the call site if it makes sense at all, rather than here.